contexts/PanelContext.tsx

```diff
@@ -1,40 +1,76 @@
 import { createContext, ReactNode, useState } from "react";
 import { PanelContentT, PanelTypeT } from "types";
 
-interface PanelContextState {
+export interface PanelContextState {
   panelType: PanelTypeT;
-  setPanelType: (panelType: PanelTypeT) => void;
-  panelContent: PanelContentT;
-  setPanelContent: (panelContent: PanelContentT) => void;
+  panelContent: PanelContentT | undefined;
   panelOpen: boolean;
-  setPanelOpen: (panelOpen: boolean) => void;
+  openPanel: (panelType: PanelTypeT, panelContent?: PanelContentT) => void;
+  closePanel: (clearPreviousPanelData?: boolean) => void;
 }
 
-const defaultPanelContextState = {
+export const defaultPanelContextState = {
   panelType: null,
-  setPanelType: () => null,
   panelContent: null,
-  setPanelContent: () => null,
   panelOpen: false,
-  setPanelOpen: () => null,
+  openPanel: () => null,
+  closePanel: () => null,
 };
 
 export const PanelContext = createContext<PanelContextState>(defaultPanelContextState);
 
 export function PanelProvider({ children }: { children: ReactNode }) {
   const [panelType, setPanelType] = useState<PanelTypeT>(null);
-  const [panelContent, setPanelContent] = useState<PanelContentT>(null);
+  const [panelContent, setPanelContent] = useState<PanelContentT | undefined>(null);
   const [panelOpen, setPanelOpen] = useState<boolean>(false);
+  const [previousPanelData, setPreviousPanelData] = useState<
+    { panelType: PanelTypeT; panelContent: PanelContentT | undefined }[]
+  >([]);
+
+  function openPanel(panelType: PanelTypeT, panelContent?: PanelContentT) {
+    setPanelType(panelType);
+    setPanelContent(panelContent);
+    setPanelOpen(true);
+    pushPanelDataOntoStack(panelType, panelContent);
+  }
+
+  function closePanel(clearPreviousPanelData = false) {
+    if (clearPreviousPanelData) {
+      setPreviousPanelData([]);
+      setPanelOpen(false);
+      return;
+    }
+    if (previousPanelData.length > 1) {
+      const previousPanel = previousPanelData[previousPanelData.length - 2];
+      setPanelType(previousPanel.panelType);
+      setPanelContent(previousPanel.panelContent);
```

Comment:
only thing id suggest here is that its awkward to have your panel stack take this combined type, then your current panel be managed by 2 separate types, why not just use the same type for current panel as `previousPanelData`, basically merging both panelType and panelContent